### PR TITLE
[MXNET-600][Scala] NDArray auto-collector

### PR DESCRIPTION
**Pull request by yzhliu**
For now user has to dispose all intermediate allocated NDArrays themselves. The `NDArrayCollector` introduced here is to provide a way for NDs to be disposed automatically.

In the test case & changes to `NDArrayIter.scala` one can find the example of how to use it.

And for Java users:
https://github.com/yzhliu/mxnet-java-example/blob/master/src/main/java/me/yzhi/mxnet/example/infer/RndImageInference.java#L28-L30
https://github.com/yzhliu/mxnet-java-example/blob/master/src/main/java/me/yzhi/mxnet/example/infer/Seq2SeqInference.java#L35-L37

In the cases above, NDArrays allocated within the scope will be disposed automatically after the code block finishes.

### CODE REVIEW

**Comment by lanking520** [comment]
Hi @yzhliu , It's been so great to see this solution. I would recommend you to apply this new feature to the `examples` in the CI for testing and that include `MNIST` and `GAN` to see if there are any more memory leaks thrown.

File: scala-package/core/src/main/scala/org/apache/mxnet/NDArrayCollector.scala [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+   * @param ndArray NDArrays need to be collected.
+   */
+  @varargs def collect(ndArray: NDArray*): Unit = {
+    currCollector.get().add(ndArray: _*)
+  }
+}
+
+class NDArrayCollector private(private val autoDispose: Boolean = true,
+                               private val doCollect: Boolean = true) {
+  private val arrays: mutable.Map[Long, NDArray] = mutable.HashMap.empty[Long, NDArray]
```
Comment (CodingCat) [inline scala-package/core/src/main/scala/org/apache/mxnet/NDArrayCollector.scala]: private variable may not need to have an explicit type

File: scala-package/core/src/main/scala/org/apache/mxnet/NDArrayCollector.scala [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+ *  }
+ *  </pre>
+ *  In the case above, the intermediate NDArrays
+ *  (created by <em>NDArray.relu</em> and <em>+</em>) will be disposed automatically. <br />
+ *  User can also decide to use dispose the collected NDArrays later: <br />
+ *  <pre>
+ *  val collector = NDArrayCollector.manual()
+ *  val res = collector.withScope {
+ *    (NDArray.relu(a) + a).toArray
+ *  }
```
[the remark this answers is not in the document]
Comment (CodingCat) [reply scala-package/core/src/main/scala/org/apache/mxnet/NDArrayCollector.scala]: if there is no other benefit, we may consider make it simpler as just a NDArray.scope, all NDArray in this scope will be automatically disposed (do not even need manual scope)...and for the other case, the user can just do what they are currently doing

**Comment by yzhliu** [comment]
@lanking520 Here you can find how we can fix the memory leak for MNIST training (using SGD): https://github.com/yzhliu/mxnet/pull/5

But I prefer to make the PR separate. You can try similar thing for GAN, and with other optimizer.

File: scala-package/core/src/main/scala/org/apache/mxnet/NDArrayCollector.scala [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+  private val arrays = mutable.HashMap.empty[Long, NDArray]
+
+  private def add(nd: NDArray*): Unit = {
+    if (doCollect) nd.foreach(arr => arrays.put(arr.handle, arr))
+  }
+
+  /**
+   * Clear the collector.
+   */
+  def clear(): Unit = {
```
[the remark this answers is not in the document]
Comment (yzhliu) [reply scala-package/core/src/main/scala/org/apache/mxnet/NDArrayCollector.scala]: When using manual(), user may want to re-use one collector:
```scala
val c = NDArrayCollector.manual()
c.withScope { ... }
...
c.clear()
c.withScope { ... }
```

File: scala-package/core/src/main/scala/org/apache/mxnet/NDArrayCollector.scala [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+   * - stored for later access (when using <em>manual</em>) <br />
+   * If the return type of scope is <em>NDArray</em> or <em>NDArrayFuncReturn</em>,
+   * it is smart enough NOT to collect or dispose the returned NDArray. <br />
+   * However in other cases, it is users' responsibility NOT to leak allocated NDArrays outside.
+   * @param body code block to be executed within the scope.
+   * @tparam T return type of the function <em>body</em>.
+   * @return The result of function <em>body</em>.
+   */
+  def withScope[T](body: => T): T = {
+    val old = NDArrayCollector.currCollector.get()
```
[the remark this answers is not in the document]
Comment (yzhliu) [reply scala-package/core/src/main/scala/org/apache/mxnet/NDArrayCollector.scala]: As we talked offline, yes.

**Comment by nswamy** [comment]
@yzhliu Thanks for elaborating with a detailed example and sorry for the delay in response. 
While I understand WithScope is what manages the deallocation of the NDArrays and using it ThreadLocal is what makes managing it possible, nevertheless my concern about thread-safety still remains. 
The use-case of producer/consumer is not very uncommon and 

```
val q = new Queue[NDArray] //assume Queue  is thread-safe

val t1 = new Thread() {
@override def run() ={
withScope {
 val a = NDArray.load()
 q.enqueue(a)
}
}
val t2 = new Thread() {
@override def run() = {
b = q.dequeue()
b.reshape()
}
}
```
This use-case or similar is not hard imagine when you are creating data pipeline and I think we should cautious in introducing new patterns to users that aren't thread-safe. 

We discussed offline about using reference-counts, any luck with that?

**Comment by yzhliu** [comment]
@nswamy unfortunately I don't think it is possible to implement within Java user code (unless we leverage finalize)  -  Users cannot access java heap (in our case,  reference number of one object) - the only way  is to dump from outside.

Moreover, not necessarily to be multi-threading, even single-thread can have problem:
```scala
var a: NDArray
collector.withScope {
a = new NDArray
}  // a is disposed here
a.toArray() // fail
```
As I stated, it is users' responsibility not to leak new NDArray outside of scope, otherwise they need to use `manual()`, and dispose it later.

Every tool has its own limitation, e.g., users are not supposed to use `HashMap` in multi-threads - otherwise they need to do synchronize themselves.

As you can see, this tool make disposing super easy: https://github.com/yzhliu/mxnet/pull/5/files#diff-19b434167a8c2d81db271f8c47975ff6R43

**Comment by nswamy** [comment]
@yzhliu I understand what you are saying and I think it is given that it doesn't work outside of the scope. 
agree every tool has its limitation, I already said that I like this :) 

Do you think changing `WithScope` to `WithThreadScope` and make it explicit would be useful for users to communicate that it is not thread-safe?

Also, I did not intend my code to be a running version, it was more for illustration of my point.. i'll call it out next time :)

**Comment by yzhliu** [comment]
@nswamy Oh no no, I was not saying your code is wrong or cannot run, I was demonstrating how to change the code a bit to safely use the collector in such situation - my code was not a runnable version either :)

As I said, even single thread can have problem, don't think it is proper to have 'Thread' in the name.

**Comment by nswamy** [comment]
you are right, changing to `ThreadScope` might again confuse people if the case you showed. 

Thanks for the clarifying about the example code, appreciate your patience :)

**Comment by yzhliu** [comment]
Thanks all people helped to review here.

**Pull request merged**